### PR TITLE
feat(authentication): enhance logging and integrate Fail2Ban configuration (#187)

### LINKED ISSUE #187: Small Feature: Improve Logging for Authentication and Proxy Scenarios
We are currently using this project and it’s been a great help! However, I would like to suggest some improvements around logging, especially when dealing with authentication and reverse proxy scenarios. Here are the main points:

1.	Reverse Proxy Handling: When using a reverse proxy, the current implementation logs the client’s IP address using r.RemoteAddr, which captures internal addresses (e.g., 172.26.0.1). Instead, we should log the actual client’s IP using the X-Forwarded-For header.

Suggested Code Change:

`// Replace r.RemoteAddr with X-Forwarded-For when behind a reverse proxy

remoteAddr := r.Header.Get("X-Forwarded-For")
if remoteAddr == "" {
    remoteAddr = r.RemoteAddr
}`

2.	Failed Login Attempts: Currently, the code logs generic login attempts, but there is no distinction between incorrect usernames and incorrect passwords. It would be more helpful to log failed attempts with specific details, such as “invalid username” or “invalid password”. Additionally, logging every login attempt can result in excessive logs, especially for failed attempts. Logging only the result (success or failure) will help reduce log volume.
Suggested Code Change:

`// Before:

zap.L().Info("login attempt", zap.String("username", username), zap.String("remote_address", r.RemoteAddr))`

`// After:

if !ok {
    zap.L().Info("invalid username", zap.String("username", username), zap.String("remote_address", remoteAddr))
    http.Error(w, "Not authorized", http.StatusUnauthorized)
    return
}

if !h.noPassword && !user.checkPassword(password) {
    zap.L().Info("invalid password", zap.String("username", username), zap.String("remote_address", remoteAddr))
    http.Error(w, "Not authorized", http.StatusUnauthorized)
    return
}`

These changes will enhance log clarity and reduce unnecessary log volume, especially for larger-scale deployments.

Thanks again for this great project! We are already using it in our environment and hope these improvements can make it even better. If you don’t have time to work on this, I’d be happy to implement these changes and submit a PR. 

Let me know your thoughts!

### PR DESCRIPTION
**Pull request by jiongxuan**
This PR includes several improvements to the authentication system's logging, along with a guide to integrating Fail2Ban to enhance the overall security of the WebDAV server.

### **Key Changes:**

1. **Improve IP Logging for Real Client IP Address:**
   - **Commit:** `feat(authentication): improve IP logging by extracting real client IP from X-Forwarded-For header`
   - Added the `getRealRemoteIP` function to correctly extract the client's real IP address when requests come through a reverse proxy.
   - Updated authentication logging to use the extracted IP instead of `r.RemoteAddr`.
   - Ensured fallback to `r.RemoteAddr` if `X-Forwarded-For` header is not present, maintaining compatibility with both proxy and non-proxy environments.

2. **Authentication Logging Enhancements:**
   - **Commit:** `feat(authentication): enhance login failure logging and reduce log volume`
   - Added logging for failed login attempts specifically indicating invalid usernames, enhancing the detail in logs.
   - Removed redundant "login attempt" logs, focusing only on final validation results, which helps reduce log volume.
   - Retained detailed logging for invalid passwords and successful authorizations for greater clarity and auditability.

3. **Fail2Ban Integration Guide:**
   - **Commit:** `docs: add Fail2Ban configuration guide to README`
   - Added a section in `README.md` to guide users on how to configure Fail2Ban with WebDAV.
   - Included examples for `filter.d` and `jail.d` configurations, allowing users to automatically ban IP addresses that exceed the allowed number of failed login attempts.

### **Issue Reference:**
This PR closes #187, which requested improvements in logging details and Fail2Ban integration for better security.

---

### **Additional Context:**

- **Real Client IP Address Extraction:**
  - The addition of `getRealRemoteIP` helps to accurately log the IP address of the actual client, especially when the server is behind a reverse proxy. This is crucial for accurate logging and security analysis.

- **Fail2Ban Configuration:**
  - The integration of Fail2Ban helps protect the WebDAV server from brute-force attacks by banning IP addresses that repeatedly fail to authenticate. This increases server resilience against malicious attempts.

- **Log Volume Optimization:**
  - Reducing log entries to only include validation outcomes (i.e., success or failure) instead of every login attempt will help maintain clean logs that are easier to monitor and analyze.

Feel free to review the changes. Let me know if any further adjustments are needed or if there are additional requirements!

### CODE REVIEW

File: lib/handler.go [hunk 1]
```diff
@@ -89,25 +89,30 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 		// Gets the correct user for this request.
 		username, password, ok := r.BasicAuth()
-		zap.L().Info("login attempt", zap.String("username", username), zap.String("remote_address", remoteAddr))
+		// Removed the login attempt log to reduce log volume and focus on final verification results
+		// zap.L().Info("login attempt", zap.String("username", username), zap.String("remote_address", remoteAddr))
```
Comment (hacdias) [inline lib/handler.go:93]: This is fine, but please remove this line. It doesn't make sense to keep it there commented.

**Review by hacdias — CHANGES_REQUESTED** [review]
Thanks for the contribution. I will merge this once the comment above is tackled.

File: lib/handler.go (hunk 1 above)
Comment (jiongxuan) [reply lib/handler.go]: Thanks for the feedback! I’ve removed the commented line as requested.

I initially left it there for comparison purposes, to highlight the change in log volume, but I agree that removing it entirely makes the code cleaner and more focused. Thanks again!

**Comment by jiongxuan** [comment]
Thanks for the feedback! I’ve removed the commented line as requested.

I initially left it there for comparison purposes, to highlight the change in log volume, but I agree that removing it entirely makes the code cleaner and more focused. Thanks again!

**Review by hacdias — APPROVED** [review]

**Pull request merged**